registry/remote/repository.go

```diff
@@ -195,7 +203,26 @@ func (r *Repository) push(ctx context.Context, expected ocispec.Descriptor, cont
 	req.ContentLength = expected.Size
 	req.Header.Set("Content-Type", expected.MediaType)
 
-	resp, err := r.client().Do(req)
+	// if the underlying client is an auth client, the content might be read
+	// more than once for obtaining the auth challenge and the actual request.
+	// To prevent double reading, the manifest is read and stored in the memory,
+	// and serve from the memory.
+	client := r.client()
+	if _, ok := client.(*auth.Client); ok && req.GetBody == nil {
+		store := cas.NewMemory()
+		err := store.Push(ctx, expected, content)
+		if err != nil {
+			return err
+		}
+		req.GetBody = func() (io.ReadCloser, error) {
+			return store.Fetch(ctx, expected)
```

Comment:
Is there any call path that would expose this to storing large chuncks in memory?

Reply:
No, this memory usage is scoped to this function and will be released after function return. Since it is only for manifests, the memory usage will be small.